### DISCUSSION TITLE
Multiple (and different) child properties in a parent object.

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3074] [Questions]
We are developing a parent object in one of our projects (Csla 4.xx).
We have four children objects, each one of different type in the parent object. 
We have been using something like this

   
             protected override void OnChildChanged(Csla.Core.ChildChangedEventArgs e)
              {
                 if (e.ChildObject is ListItems)
                 {
                      BusinessRules.CheckRules(ListItemsProperty);
                      OnPropertyChanged(ListItemsProperty);
                  }
                        base.OnChildChanged(e);
              }
   

We are thinking of using the code below but it just doesn't look elegant : )

             protected override void OnChildChanged(Csla.Core.ChildChangedEventArgs e)
              {
                 if (e.ChildObject is ChildType1)
                 {
                      BusinessRules.CheckRules(ChildType1Property);
                      OnPropertyChanged(ChildType1Property);
                  }

               if (e.ChildObject is ChildType2)
                 {
                      BusinessRules.CheckRules(ChildType2Property);
                      OnPropertyChanged(ChildType2Property);
                  }
                base.OnChildChanged(e);
              }

      

King Regards

**Comment by rockfordlhotka** [comment]
I wonder if you could create an interface for these child types:

```c#
public interface IAmAChild
{
  IPropertyInfo ParentProperty { get; }
}
```

Then the child types would implement that interface, using the `Parent` property to reach up to the parent to find the `IPropertyInfo` registered for the specific child object.

Just a thought, but it might make your `OnChildChanged` very simple.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka  I am afraid I am not following, how will the OnChildChanged look like or are we going to remove OnChildChanged from the parent object?

Kind Regards

**Reply by rockfordlhotka** [reply]
I'm sitting in an auto dealer while my truck's oil is changed, so this is just ideas.

The parent would have this code:

```c#
         protected override void OnChildChanged(Csla.Core.ChildChangedEventArgs e)
          {
             if (e.ChildObject is IAmAChild child)
             {
                  BusinessRules.CheckRules(child.ParentProperty);
                  OnPropertyChanged(child.ParentProperty);
              }
            base.OnChildChanged(e);
          }
```

The child class would implement `IAmAChild` like this:

```c#
  IPropertyInfo IAmAChild.ParentProperty
  {
    get
    {
      var manager = Parent.FieldManager;
      // use manager to find the FieldDataInfo object referncing 'this'
      // use manager to find the IPropertyInfo for that FieldDataInfo object
    }
  }
```

Though as I type that, it does occur to me that you might not be able to get at the field manager via the `Parent` property - I'm not sure if there's a way to do that.

If not, you might need to actually implement a `GetPropertyInfoForChild(object child)` method on the parent.

```c#
  internal IPropertyInfo GetPropertyInfoForChild(object child)
  {
      // use FieldManager to find the FieldDataInfo object referncing 'this'
      // use manager to find the IPropertyInfo for that FieldDataInfo object
  }
```

**Reply by rockfordlhotka** [reply]
Whether any of this is more elegant or not is probably debatable at this point :)

Unless you use custom base types, in which case all this could go in the base types, and then I do think it would be elegant.

**Reply by Chicagoan2016** [reply]
@rockfordlhotka You Sir, saved the day in two ways : )
First of all, thank you for the awesome code (as always)! We have custom base classes and I am going to give it a try.
secondly, I just remembered my car needs oil change, although the onboard monitoring system tells me 27% oil life is remaining but it's already more than 5k miles since the last oil change.
Regards

**Reply by swegele** [reply]
Give an update on if and how it works out for you.  I'm curious.

**Reply by Chicagoan2016** [reply]
@swegele I tried but I couldn't get it to work, I had to implement that 'use case' in a hurry and ended up using the 'redundant' code (as shown in my question)
